internal/util/imagefetcher/eagerfetcher.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package imagefetcher
+
+import (
+	"fmt"
+	"os/exec"
+	"sync"
+
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// Filter wraps another container-based kio.Filter. It starts to fetch the
+// required docker image in a goroutine in the background when it is created.
+// The underlying filter will be executed only after the image fetching
+// goroutine completes.
+type Filter struct {
+	image string
+	fltr  kio.Filter
+	wg    *sync.WaitGroup
+}
+
+// NewFilter creates a kio.Filter that wraps another container-based kio.Filter.
+func NewFilter(img string, fltr kio.Filter) *Filter {
+	var wg sync.WaitGroup
+	f := &Filter{
+		image: img,
+		fltr:  fltr,
+		wg:    &wg,
+	}
+	wg.Add(1)
+	go f.fetchImage()
```

Comment:
Doing this as part of this constructor is making a bit uncomfortable because any sort of processing is not expected while creating the filter itself.
An alternative will be introducing a new interface (`preloadable`) with `PreLoad()` method. And invoking `PreLoad()` before pipeline.execute on all the filters if they are preloadable. But it's making caller code aware of the caching stuff. Not great either IMO.  

So may be, it's not that bad.